### ISSUE TITLE
AWS Console shows Agent as disconnected, but no errors in logs and agent is running

### ISSUE THREAD

**Issue by bodnarbm** [aws/amazon-ecs-agent#577]
According to the ECS Console one of our Container Instances are showing as Agent Connected: False, but when we check the box, the ecs-agent docker container is running and the logs don't seem to indicate any error related to the agent connecting (other than an error about lack of metrics) and seem to indicate that the agent is communicating with the cluster.

![screen shot 2016-10-31 at 9 15 37 am](https://cloud.githubusercontent.com/assets/121035/19855380/a27489f6-9f4a-11e6-9e80-f60238d4125c.png)

This seems to cause no new tasks to be scheduled for that container instance, placing a higher burden on the other instances in the cluster.

Attached are the two most recent ecs-agent logs.
[ecs-agent.log.2016-10-31-12.txt](https://github.com/aws/amazon-ecs-agent/files/561807/ecs-agent.log.2016-10-31-12.txt)
[ecs-agent.log.2016-10-31-13.txt](https://github.com/aws/amazon-ecs-agent/files/561809/ecs-agent.log.2016-10-31-13.txt)

**Comment by richardpen** [comment]
@bodnarbm On first investigation we were not able to find the reason why the agent isn't connected. The provided logs doesn't contain any useful information for us to find the root cuase, could you provide the following information to help us find the root cause, you can put it here or send it directly to me at: penyin (at) amazon.com:
* Run command `docker kill -s USR1 ecs-agent` and then send us the whole agent logs(in /var/log/ecs/).

And after that, can you restart the agent by running `sudo stop ecs` and `sudo start ecs` to see if you can get ride of this issue.

thanks,
peng

**Comment by bodnarbm** [comment]
Thanks @richardpen. I ran `docker kill -s USR1 ecs-agent` and gathered up all my logs. I just emailed them to you.

Afterwards, I restarted the ecs-agent container and that Container Instance now reports as connected to the cluster. I will monitor if it happens again. Otherwise, I hope the logs I emailed can provide some clue as to what happened.

Also, in case it helps, we are actually not running the ecs optimized ami, but are running the ecs-agent container on an ubuntu installation.

**Comment by richardpen** [comment]
@bodnarbm thanks for your information, I took a look at your logs and found some issue with the agent, I have created the PR #580 to fix it. But that may not be the root cause of your problem, and the logs provided only contains the information after the agent is already disconnected, could you provide me the logs as I mentioned above when this issue happened again, actually I'm interested in the logs when the issue happened? It would also be helpful if you can run ecs-agent in debug mode with `--env= ECS_LOGLEVEL=debug`.

thanks,
Peng

**Comment by bodnarbm** [comment]
No problem. We will watch for an agent disconnecting again and try to grab the logs as quickly as possible.

**Comment by bodnarbm** [comment]
I set up a monitoring script that checks every 5 mins on whether the agent is connected using the aws cli and that alerts me to the disconnect. We had an alert go off near the end of last hour. I just emailed you the log. The disconnect would have occurred sometime in the five minutes before 7:52 pm eastern.

You will see the agent restart sometime around 7:52. This is because our monitoring script kills the container and restarts it if it detects the disconnected. This is a temporary workaround to resolve the above issue of tasks not scheduling on the Container Instance during the persistent disconnect status.

**Comment by richardpen** [comment]
@bodnarbm Thanks for providing the logs, I found in the logs the agent disconnected at:
```
2016-11-02T23:51:47Z [WARN] ACS Connection hasn't had any activity for too long; closing connection
```
And it was restarted at:
```
2016-11-02T23:52:06Z [INFO] Starting Agent: Amazon ECS Agent - v1.13.0 (aebcbca)
```
I should have pointed out earlier, the agent will disconnect and connect periodically if there is no activity for some time(this happened per the log above), and the disconnection duration should be less than 2 minutes. So if you continue seeing the disconnection from console for 2 minutes, there may be a problem, otherwise the agent is working in desired way. 

While in the previous log, the agent was disconnected for more than 20 hours, which isn't the normal case, there may be a problem in agent, we will continue debugging the root cause. At the mean time, I think you can modify your script to monitor the disconnection for ~2 minutes and then restart it to cooperate the way agent works.

thanks,
Peng

**Comment by bodnarbm** [comment]
No problem. I'll make the modification to the monitoring script and will update if we see the problem again with disconnects lasting longer than anticipated.

**Comment by bodnarbm** [comment]
@richardpen I just sent you some updated logs from over the weekend. Unfortunately I think they logs may have rolled off the ones from when the issue seemed to appear again this weekend. I hope there is some useful information in the logs.

Looking at cloudwatch, it seems the agent stopped reporting metrics to sometime around 2016-11-05@0100.

![screen shot 2016-11-07 at 8 33 07 am](https://cloud.githubusercontent.com/assets/121035/20059555/d741971c-a4c4-11e6-94d6-3d129e2cf8b7.png)

**Comment by richardpen** [comment]
@bodnarbm Thanks for providing the logs. The metrics gap between 11/05~11/07 is caused by the same reason as #532, the fix will be available in our next release. For the acs disconnection we will still work on finding the root cause.

Thanks.

**Comment by richardpen** [comment]
@bodnarbm We released [v1.13.1](https://github.com/aws/amazon-ecs-agent/releases/tag/v1.13.1) today, which has several bug fix included. Could you try the latest version? I'm closing this for now, feel free to reopen it if you still run into this issue in the future.
